keras/backend/jax/distribution_lib.py

```diff
@@ -43,6 +43,13 @@ def distribute_value(value, tensor_layout):
     return jax.device_put(value, tensor_layout)
 
 
+def relayout(value, tensor_layout):
```

Comment:
Indeed. I think the major difference here is one of them is suppose to work in jitted function, and the other one works in eager.  Currently I don't think we have a good way to detect what mode is user code in, and auto choose the proper API for them.

I think the jax `with_sharding_constraint` has a good indication that it applies the constraint to a intermediate tensor/state within the function. Maybe we should just aligned with that?

Reply:
Q: what's the difference between `with_sharding_constraint` applied to the input tensor (i.e. the data array), vs calling `device_put` on the input tensor? Right now we use `device_put` for data distribution. Does `with_sharding_constraint` work for that?

Reply:
So I have a some chat in the jax user group, and the conclusion is that the `with_sharding_constraint` is designed to be used only in jitted function. It might not work in a lot of cases outside of jax.jit.

I think we should rely on the device_put in the pure eager context, for input data, as well as variable initialization.

Reply:
Let's do this, I think:

- Have a separate `distribute_variable` and `distribute_tensor` API. Apparently this may be needed for TF?
- In JAX `distribute_tensor`, check if we're in a tracing context. If so, use `sharding_constraint`. If not, use `device_put`.

Reply:
Done. I have added at check for distribute_tensor for whether its in the jitted context, and I think it might not be a cheap check if we have to do it very often. I also didn't find a proper way to do this kind of check https://github.com/google/jax/discussions/9241. We might want to check with Jax team for this.